File: 1. Creational Design Patterns/Factory Method/C#/FactoryMethodExample.cs
// Factory Method Pattern Example in C#
// Scenario: Logistics creates appropriate Transport (Truck or Ship) at runtime.
// Compile: csc FactoryMethodExample.cs && FactoryMethodExample.exe [sea]

using System;

namespace FactoryMethodDemo
{
    interface ITransport
    {
        void Deliver();
    }

    class Truck : ITransport
    {
        public void Deliver() => Console.WriteLine("Deliver by land in a box.");
    }

    class Ship : ITransport
    {
        public void Deliver() => Console.WriteLine("Deliver by sea in a container.");
    }

    abstract class Logistics
    {
        public void PlanDelivery()
        {
            var transport = CreateTransport();
            transport.Deliver();
        }
        protected abstract ITransport CreateTransport();
    }

    class RoadLogistics : Logistics
    {
        protected override ITransport CreateTransport() => new Truck();
    }

    class SeaLogistics : Logistics
    {
        protected override ITransport CreateTransport() => new Ship();
    }

    class Program
    {
        static void Main(string[] args)
        {
            Logistics logistics = (args.Length > 0 && args[0].Equals("sea", StringComparison.OrdinalIgnoreCase))
                ? new SeaLogistics()
                : new RoadLogistics();
            logistics.PlanDelivery();
        }
    }
}

File: 1. Creational Design Patterns/Factory Method/Java/FactoryMethodExample.java
// Factory Method – Java Example
interface Transport {
    void deliver();
}

class Truck implements Transport {
    public void deliver() {
        System.out.println("Deliver by land in a box.");
    }
}

class Ship implements Transport {
    public void deliver() {
        System.out.println("Deliver by sea in a container.");
    }
}

abstract class Logistics {
    void planDelivery() {
        Transport t = createTransport();
        t.deliver();
    }

    abstract Transport createTransport();
}

class RoadLogistics extends Logistics {
    Transport createTransport() { return new Truck(); }
}

class SeaLogistics extends Logistics {
    Transport createTransport() { return new Ship(); }
}

public class FactoryMethodExample {
    public static void main(String[] args) {
        Logistics log = (args.length > 0 && args[0].equals("sea"))
                ? new SeaLogistics()
                : new RoadLogistics();
        log.planDelivery();
    }
}
